### PR TITLE
impr: crunch like way of doing run

### PR DESCRIPTION
**Pull request by AntoineToussaint**

### CODE REVIEW

File: endersgame/attackers/__init__.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+UP = 1.
+DOWN = -1.
+NONE = 0.
```
Comment (Caceresenzo) [inline endersgame/attackers/__init__.py:3]: todo: do not use floats

File: endersgame/crunch/__init__.py [hunk 2]
```diff
@@ -0,0 +1,3 @@
+from endersgame.crunch.univariate import Crunch
+
+crunch = Crunch()
```
Comment (Caceresenzo) [inline endersgame/crunch/__init__.py:3]: question: why?

File: endersgame/crunch/replay.py [hunk 3]
```diff
@@ -52,14 +54,24 @@ def filter(self, pt: MarketDataPoint):
...
 
-    def tick(self) -> StreamBatch:
+    def values(self, delay: float) -> Iterator[StreamPoint]:
+        while self.idx < len(self.timestamps):
+            batch = self.tick()
+            for pt in batch.points:
+                if pt.substream_id not in self.substream_ids_only:
+                    continue
+                yield pt
+                time.sleep(delay)
```
Comment (Caceresenzo) [inline endersgame/crunch/replay.py:64]: todo: replay should not sleep and just run as fast as possible

Or enable sleeping with a flag.

File: endersgame/crunch/streams.py [hunk 4]
```diff
@@ -1,18 +1,18 @@
...
 
 from pydantic import BaseModel
 
 
 class Prediction(BaseModel):
     value: float
-    # TODO: Oneof?
-    t: datetime = None # absolute time
+    t: Optional[datetime] = None # absolute time
     n: int = None # relative horizon
```
Comment (Caceresenzo) [inline endersgame/crunch/streams.py:10]: todo: single letter variable name are forbidden

File: endersgame/crunch/streams.py [hunk 5]
```diff
@@ -1,18 +1,18 @@
...
 
 class StreamPoint(BaseModel):
-    substream_id: str
+    substream_id: Optional[str]
     value: float
-    t: datetime = None # absolute time
+    t: Optional[datetime] = None # absolute time
     n: int = None      # relative time
```
Comment (Caceresenzo) [inline endersgame/crunch/streams.py:16]: todo: single letter variable name are forbidden

**Review by Caceresenzo — CHANGES_REQUESTED** [review]

File: endersgame/attackers/__init__.py (hunk 1 above)
Comment (microprediction) [reply endersgame/attackers/__init__.py:3]: Actually we should encourage floats. See https://github.com/microprediction/endersgame/blob/main/endersgame/attackers/calibratedattacker.py

**Pull request closed without merging**